networkit/cpp/centrality/ComplexPaths.cpp

```diff
@@ -0,0 +1,205 @@
+ *            https://github.com/drguilbe/complexpaths.git
+ *
+ *  see [ Guilbeault, D., Centola, D. Topological measures for
+ *        identifying and predicting the spread of complex contagions.
+ *        Nat Commun 12, 4430 (2021).
+ *        https://doi.org/10.1038/s41467-021-24704-6 ]
+ *
+ */
+
+#include <algorithm>
+#include <iostream>
+#include <iterator>
+#include <random>
+#include <vector>
+
+#include <networkit/auxiliary/Random.hpp>
+#include <networkit/base/Algorithm.hpp>
+#include <networkit/centrality/ComplexPaths.hpp>
+#include <networkit/distance/BFS.hpp>
+#include <networkit/distance/PrunedLandmarkLabeling.hpp>
+#include <networkit/graph/Graph.hpp>
+
+namespace NetworKit {
+
+ComplexPathAlgorithm::ComplexPathAlgorithm(const Graph &G, count threshold, Mode mode, node start)
+    : inputGraph{&G}, mode{mode}, start{start}, threshold{threshold}, normPaths{false} {
+    if (mode == Mode::singleNode && start == none)
+        throw std::runtime_error("complexPathAlgorithm: provide node != none in Mode::singleNode.");
+    if (mode == Mode::allNodes && start != none)
+        throw std::runtime_error("complexPathAlgorithm: start node ignored in Mode::allNodes.");
+}
+
+void ComplexPathAlgorithm::normalize() {
+    if (mode != Mode::allNodes) {
+        WARN("complexPathAlgorithm: no normalization in Mode::singleNode.");
+        return;
+    }
+    normPaths = true;
+}
+
+void ComplexPathAlgorithm::run() {
+    if (mode == Mode::singleNode) {
+        complexPathGraph = complexPathsGraph(start, threshold, &adopters);
+    } else {
+        complexPathsLengths = complexPathLength(threshold);
+    }
+    hasRun = true;
+}
+
+std::vector<double> ComplexPathAlgorithm::getPLci() {
+    if (mode != Mode::allNodes)
+        WARN("complexPathAlgorithm[getPLci]: no results in Mode::singleNode.");
+    assureFinished();
+    return complexPathsLengths;
+}
+
+Graph ComplexPathAlgorithm::getComplexGraph() {
+    if (mode != Mode::singleNode)
+        WARN("complexPathAlgorithm[getComplexGraph]: no results in Mode::allNodes.");
+    assureFinished();
+    return complexPathGraph;
+}
+
+std::vector<node> ComplexPathAlgorithm::getAdopters() {
+    if (mode != Mode::singleNode)
+        WARN("complexPathAlgorithm[getAdopters]: no results in Mode::allNodes.");
+    assureFinished();
+    return adopters;
+}
+
+static void addNewEdge(Graph &g, node u, node v) {
+    if (g.hasEdge(u, v))
+        return;
+    g.addEdge(u, v);
+}
+
+std::vector<node> ComplexPathAlgorithm::generateSeeds(node seed, const Graph &g, count threshold) {
+    // activate threshold-1 random neighbors of seed:
+    auto seedsNeeded = threshold - 1;
+
+    std::vector<node> seeds;
+    seeds.reserve(4);
+
+    auto neighbors = Graph::NeighborRange<>(g, seed);
+
+    auto randomGen = Aux::Random::getURNG();
+    std::sample(neighbors.begin(), neighbors.end(), std::back_inserter(seeds), seedsNeeded,
+                randomGen);
+    auto needMore = seedsNeeded - seeds.size();
```

Comment:
If `seeds.size()` is bigger than `seedsNeeded` this will overflow.
```suggestion
    if (seeds.size() >= seedsNeeded) return seeds;
    const count additionalSeeds = seedsNeeded - seeds.size();
```